### PR TITLE
Misc: Serde serialize/deserialize support + Refactor PoseidonHash primitive and gadget code + MSM Affine BugFix + other minor functions/gadgets

### PR DESCRIPTION
**Pull request by 95DDB**

### CODE REVIEW

File: algebra/src/curves/mod.rs [hunk 1]
```diff
@@ -240,6 +254,28 @@ pub trait AffineCurve:
     #[must_use]
     fn is_zero(&self) -> bool;
 
+    /// Returns a group element if the set of bytes forms a valid group element,
+    /// otherwise returns None. This function is primarily intended for sampling
+    /// random group elements from a hash-function or RNG output.
```
Comment (UlrichHaboeck75) [inline algebra/src/curves/mod.rs:259]: Very misleading comment. I'd write "Tries to sample a group element...."

File: algebra/src/curves/models/short_weierstrass_jacobian.rs [hunk 2]
```diff
@@ -314,6 +319,72 @@ impl<P: Parameters> FromBytesChecked for GroupAffine<P> {
...
+                //Attempt to get the y coordinate from its parity and x
+                match Self::get_point_from_x_and_parity(x, parity_flag_set) {
+
+                    //Check p belongs to the subgroup we expect
+                    Some(p) => {
+                        if p.is_in_correct_subgroup_assuming_on_curve() {
+                            Ok(p)
+                        }
+                        else {
+                            let e = BitSerializationError::NotPrimeOrder;
```
Comment (UlrichHaboeck75) [inline algebra/src/curves/models/short_weierstrass_jacobian.rs]: As this implementation is for a generic affine group, it is not assured that p is the only prime factor in the number of points of the curve, nor that the prime order p has multiplicity 1. I'd through a different error, if possible.

File: algebra/src/fft/domain/mod.rs [hunk 3]
```diff
@@ -168,12 +173,38 @@ pub trait EvaluationDomain<F: PrimeField>: Debug
...
+                *tau_minus_r *= l;
             });
             u
         }
     }
 
+    /// Given an index which assumes the first elements of this domain are the elements of
+    /// another (sub)domain with size size_s,
+    /// this returns the actual index into this domain.
+    fn reindex_by_subdomain(&self, other_size: usize, index: usize) -> usize {
```
Comment (UlrichHaboeck75) [inline algebra/src/fft/domain/mod.rs:185]: As pointed out in [my comment](https://github.com/HorizenOfficial/ginger-lib/pull/34#discussion_r592218326) for the  FFT pull request, this function is not very elegant.

File: algebra/src/fields/arithmetic.rs [hunk 4]
```diff
@@ -439,18 +437,6 @@ macro_rules! impl_multiplicative_ops_from_ref {
             }
         }
 
-        #[allow(unused_qualifications)]
-        impl<P: $params> MulShort<Self> for $type<P> {
```
Comment (UlrichHaboeck75) [inline algebra/src/fields/arithmetic.rs:443]: Why are the macros in arithmetic.rs, but the arithemtics in macros.rs?

File: algebra/src/fields/models/cubic_extension.rs [hunk 5]
```diff
@@ -462,6 +464,9 @@ impl<'a, P: CubicExtParameters> SubAssign<&'a Self> for CubicExtField<P> {
     }
 }
 
+impl_additive_ops_from_ref!(CubicExtField, CubicExtParameters);
+impl_multiplicative_ops_from_ref!(CubicExtField, CubicExtParameters);
```
Comment (UlrichHaboeck75) [inline algebra/src/fields/models/cubic_extension.rs:468]: I'd moved that after the implementation of mul_assign().

File: algebra/src/lib.rs [hunk 6]
```diff
@@ -50,4 +50,49 @@ pub mod fft;
 #[cfg(feature = "fft")]
 pub use self::fft::*;
 
-pub type Error = Box<dyn std::error::Error>;
\ No newline at end of file
+pub type Error = Box<dyn std::error::Error>;
+
+/// Returns the base-2 logarithm of `x`.
```
Comment (UlrichHaboeck75) [inline algebra/src/lib.rs]: returns ceil(log_2(x)).

File: primitives/src/crh/poseidon/batched_crh.rs [hunk 7]
```diff
@@ -1,19 +1,26 @@
...
-impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr=F>> PoseidonBatchHash<F, P> {
+impl<F, P, SB> PoseidonBatchHash<F, P, SB>
+    where
+        F: PrimeField,
+        P: PoseidonParameters<Fr = F>,
+        SB: PoseidonBatchSBox<P>,
+{
 
-    fn poseidon_full_round(vec_state: &mut [Vec<P::Fr>], round_cst_idx: &mut usize) {
+    fn poseidon_full_round(vec_state: &mut [Vec<P::Fr>], round_cst_idx: &mut usize, last: bool) {
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/batched_crh.rs]: The name of this function conflicts the usual names in the context of Poseidon. 
Every round consists of an Add_constants layer, an (partial/full) S-Box layer, and the Matrix-Mix layer. 
I'd separate the adding of the constants from that function, and rebrand it into `fn sbox_layer_full()`. 
The same for the `poseidon_partial_round()` below.

Comment (UlrichHaboeck75) [reply primitives/src/crh/poseidon/batched_crh.rs]: Besides that, we will need no boolean to distinguish normal full rounds from the last. 
But I'd expect this to be part of a future PR.

File: primitives/src/crh/poseidon/parameters/bn382.rs [hunk 8]
```diff
@@ -0,0 +1,484 @@
+use crate::crh::{
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/parameters/bn382.rs:1]: Just for completeness: the parameters here are probably not reproducible, beside that the claimed generation script is not the updated one.

File: primitives/src/crh/poseidon/sbox.rs [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+    const MDS_CST_SHORT: &'static [Self::Fr];  // The MDS matrix for fast matrix multiplication
+}
+
+pub trait PoseidonSBox<P: PoseidonParameters>: SBox<Field = P::Fr, Parameters = P> {
+
+    // Function that does the scalar multiplication
+    fn scalar_mul(res: &mut <Self as SBox>::Field, state: &mut [<Self as SBox>::Field], start_idx_cst: usize);
+
+    // Function that does the mix matrix
+    fn matrix_mix(state: &mut Vec<<Self as SBox>::Field>)
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/sbox.rs]: As mentioned above: It is unusual to consider the matrix_mix as part of the S-Box layer.

File: primitives/src/crh/poseidon/sbox.rs [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    _parameters: PhantomData<P>,
+}
+
+impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonInverseParameters<Fr = F>> SBox for PoseidonInverseSBox<F, P> {
+    type Field = F;
+    type Parameters = P;
+
+    #[inline]
+    fn apply_full(state: &mut Vec<F>, last: bool) {
+        // Apply Montomgery's simulateneous inversion
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/sbox.rs]: Let's just call that batch inversion.

File: primitives/src/crh/poseidon/sbox.rs [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+
+            let z_2 = w_bar * &w2;
+            w_bar = w_bar * &state[2];
+            state[2] = z_2;
+            let z_1 = w_bar * &state[0];
+            state[0] = w_bar * &state[1];
+            state[1] = z_1;
+        }
+
+        if !last {
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/sbox.rs]: As mentioned before: no separate treatment in last round.

File: primitives/src/crh/poseidon/sbox.rs [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+        // Apply the matrix mix
+        Self::matrix_mix(state);
+    }
+}
+
+impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonInverseParameters<Fr = F>> PoseidonSBox<P> for PoseidonInverseSBox<F, P> {
+
+    // It uses a partial Montgomery multiplication defined as PM(x, t) = x * t * 2^-64 mod M
+    // t is a 64-bit matrix constant. In the algorithm, the constants are represented in
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/sbox.rs]: Misleading comment. 
PM(x,t) in this notation takes x and t as Montgomery representation with respect to the "long" and "short" Montgomery constants R_1 and R_2 = 2^64, respectively? If so, the representation t itself is 64 bit long, not t*2^64 as claimed below.

File: primitives/src/crh/poseidon/sbox.rs [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+                    vec_state[i][j] = w_bar * &w[idx as usize];
+                    w_bar = w_bar * &vec_1;
+                    idx -= 1;
+                }
+            }
+        }
+
+        if !last {
+            // Perform the matrix mix
+            for i in 0..vec_state.len() {
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/sbox.rs]: Isn't there any better way to perform the matrix multiplication in parallel?

File: primitives/src/crh/poseidon/sbox.rs [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+        let w1 = *elem * &(*elem);
+        let w2 = w1 * &w1;
+        *elem = w2 * &(*elem);
+    }
+}
+
+impl<F: PrimeField, P: PoseidonParameters<Fr = F>> PoseidonSBox<P> for PoseidonQuinticSBox<F, P> {
+    // It uses Montgomery multiplication
+    // Constants are defined such that the result is x * t * 2^n mod M,
+    // that is the Montgomery representation of the operand x * t mod M, and t is the 64-bit constant
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/sbox.rs]: This comment is wrong/misleading, in particular in regard of the above comment.
I'd simply write "Uses short Montgomery multiplication" or use the explanation from above.

File: primitives/src/crh/poseidon/sbox.rs [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+    type Parameters = P;
+
+    #[inline]
+    fn apply_full(state: &mut Vec<F>, last: bool) {
+        // Apply the S-BOX to each of the elements of the state vector
+        for i in 0..P::T {
+            Self::exp_by_5(&mut state[i]);
+        }
+
+        if !last {
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/sbox.rs]: no separate treatment in the last round.

File: primitives/src/merkle_tree/field_based_mht/parameters/bn382.rs [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+        9090604893589195550,
+        1731006251955805765,
+        1348662098166494253,
+        8636612641322414294,
+        1080867906211844884
+     ]));
+
+pub const BN382_MHT_POSEIDON_PARAMETERS: FieldBasedMerkleTreePrecomputedEmptyConstants<'static, BN382FrPoseidonHash> =
+    FieldBasedMerkleTreePrecomputedEmptyConstants {
+        nodes: &[
```
Comment (UlrichHaboeck75) [inline primitives/src/merkle_tree/field_based_mht/parameters/bn382.rs:25]: How are these constants determined?

File: proof-systems/src/groth16/benches/gro16_bench.rs [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+
+impl<F: PrimeField> ConstraintSynthesizer<F> for Benchmark<F> {
+    fn generate_constraints<CS: ConstraintSystem<F>>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(), SynthesisError> {
+        assert!(self.inputs.len() >= 2);
+        assert!(self.num_constraints >= self.inputs.len());
+
+        let mut variables: Vec<_> = Vec::with_capacity(self.inputs.len());
```
Comment (UlrichHaboeck75) [inline proof-systems/src/groth16/benches/gro16_bench.rs]: Why allocating just inputs.len() many variables, as we know already how many variables the circuit has in advance?

File: proof-systems/src/groth16/benches/gro16_bench.rs [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+use criterion::Criterion;
+
+use std::marker::PhantomData;
+
+pub struct Benchmark<F: PrimeField> {
+    inputs: Vec<Option<F>>,
+    num_constraints: usize,
+}
+
+impl<F: PrimeField> ConstraintSynthesizer<F> for Benchmark<F> {
```
Comment (UlrichHaboeck75) [inline proof-systems/src/groth16/benches/gro16_bench.rs:26]: Let's add a comment on the specifics of this test circuit. 
For instance that has R1CS density 1, and keeps the synthesizer cost low (no field inversions).

File: proof-systems/src/groth16/benches/gro16_bench.rs [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+        Ok(())
+    }
+}
+
+pub struct BenchmarkHighDensities<F: Field> {
+    num_constraints: usize,
+    _engine:         PhantomData<F>,
+}
+
+impl<F: Field> ConstraintSynthesizer<F> for BenchmarkHighDensities<F> {
```
Comment (UlrichHaboeck75) [inline proof-systems/src/groth16/benches/gro16_bench.rs:72]: The same here. Please let us add comments!

File: r1cs/gadgets/crypto/src/crh/poseidon/mod.rs [hunk 20]
```diff
@@ -158,7 +135,7 @@ impl<
         {
             // Apply the S-BOX to each of the elements of the state vector
             for (j, d) in state.iter_mut().enumerate() {
-                Self::mod_inv_sbox(cs.ns(|| format!("mod_inv_S-Box_4_{}_{}", P::R_F-1, j)), d)?;
+                SBG::apply(cs.ns(|| format!("mod_inv_S-Box_4_{}_{}", P::R_F-1, j)), d)?;
```
Comment (UlrichHaboeck75) [inline r1cs/gadgets/crypto/src/crh/poseidon/mod.rs]: As stated before, the last round is not treated differently than the other full rounds.

File: r1cs/gadgets/crypto/src/merkle_tree/mod.rs [hunk 21]
```diff
@@ -1,17 +1,85 @@
...
-use crate::FixedLengthCRHGadget;
+use primitives::{crh::FixedLengthCRH, merkle_tree::*, FieldBasedHash};
+use crate::{FixedLengthCRHGadget, FieldBasedHashGadget};
 
 use std::borrow::Borrow;
+use r1cs_std::fields::fp::FpGadget;
 
 pub mod field_based_mht;
 
+pub trait FieldBasedMerkleTreePathGadget<
```
Comment (UlrichHaboeck75) [inline r1cs/gadgets/crypto/src/merkle_tree/mod.rs:13]: Why do we introduce a generic FieldBasedMerkleTreePathGadget beside the binary one? Do we ever target a different arity?

**Review by UlrichHaboeck75 — CHANGES_REQUESTED** [review]
Beside some minor things addressed in my comments, the main issue is to restructure the PoseidonHash low level functions to be in alignment with the usual names (S-Box does not involve the linear layer, etc.).

File: primitives/src/crh/poseidon/batched_crh.rs (hunk 7 above)
Comment (95DDB) [reply primitives/src/crh/poseidon/batched_crh.rs]: Done

File: primitives/src/crh/poseidon/sbox.rs (hunk 9 above)
Comment (95DDB) [reply primitives/src/crh/poseidon/sbox.rs]: Done

File: r1cs/gadgets/crypto/src/merkle_tree/mod.rs (hunk 21 above)
Comment (95DDB) [reply r1cs/gadgets/crypto/src/merkle_tree/mod.rs:13]: We may in future, yes

File: primitives/src/crh/poseidon/parameters/bn382.rs (hunk 8 above)
Comment (95DDB) [reply primitives/src/crh/poseidon/parameters/bn382.rs:1]: Re-generated using yours

File: primitives/src/merkle_tree/field_based_mht/parameters/bn382.rs (hunk 16 above)
Comment (95DDB) [reply primitives/src/merkle_tree/field_based_mht/parameters/bn382.rs:25]: Re-generated using update BN382 Poseidon constants. They are simply the hash of empty nodes up to a certain height

File: algebra/src/fft/domain/mod.rs (hunk 3 above)
Comment (95DDB) [reply algebra/src/fft/domain/mod.rs:185]: Open an issue please, we will address this in future

File: algebra/src/curves/models/short_weierstrass_jacobian.rs [hunk 22]
```diff
@@ -314,6 +319,72 @@ impl<P: Parameters> FromBytesChecked for GroupAffine<P> {
...
+        let p = if self.infinity {P::BaseField::zero()} else {self.x};
+        let mut res = p.write_bits();
+
+        // Is this the point at infinity? If so, set the most significant bit.
+        res.push(self.infinity);
+
+        // Is the y-coordinate the odd one of the two associated with the
+        // x-coordinate? If so, set the third-most significant bit so long as this is not
+        // the point at infinity.
+
```
[the remark this answers is not in the document]
Comment (95DDB) [reply algebra/src/curves/models/short_weierstrass_jacobian.rs]: Fixed

File: algebra/src/curves/models/short_weierstrass_jacobian.rs (hunk 2 above)
Comment (95DDB) [reply algebra/src/curves/models/short_weierstrass_jacobian.rs]: Introduced a new error "NotInTheCorrectSubgroup", just to be more generic

File: algebra/src/lib.rs (hunk 6 above)
Comment (95DDB) [reply algebra/src/lib.rs]: Done

File: primitives/src/crh/poseidon/sbox.rs (hunk 10 above)
Comment (95DDB) [reply primitives/src/crh/poseidon/sbox.rs]: Done

File: primitives/src/crh/poseidon/sbox.rs (hunk 12 above)
Comment (95DDB) [reply primitives/src/crh/poseidon/sbox.rs]: Done

File: primitives/src/crh/poseidon/sbox.rs (hunk 14 above)
Comment (95DDB) [reply primitives/src/crh/poseidon/sbox.rs]: Done

File: proof-systems/src/groth16/benches/gro16_bench.rs (hunk 17 above)
Comment (95DDB) [reply proof-systems/src/groth16/benches/gro16_bench.rs]: Pre-allocation size should be inputs.len() + num_constraints, fixed

File: proof-systems/src/groth16/benches/gro16_bench.rs (hunk 18 above)
Comment (95DDB) [reply proof-systems/src/groth16/benches/gro16_bench.rs:26]: Done

File: primitives/src/crh/poseidon/batched_crh.rs [hunk 23]
```diff
@@ -1,126 +1,99 @@
...
+use algebra::PrimeField;
 use std::marker::PhantomData;
 use crate::crh::BatchFieldBasedHash;
-use crate::{Error, PoseidonParameters, matrix_mix_short, PoseidonHash};
+use crate::{Error, PoseidonParameters, PoseidonHash, BatchSBox};
+use rayon::prelude::*;
 
-/// Batch evaluation for x^{-1}-Poseidon with MDS matrix supporting short Montgomery multiplication.
-pub struct PoseidonBatchHash<F: PrimeField, P: PoseidonParameters<Fr = F>>{
+pub struct PoseidonBatchHash<F: PrimeField, P: PoseidonParameters<Fr = F>, SB: BatchSBox<Field = F, Parameters = P>>
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/batched_crh.rs:7]: Is it really worth to do introduce  an extra trait BatchSBox? 
I wouldn't know of any other than the inversion SBox.

File: primitives/src/crh/poseidon/batched_crh.rs [hunk 24]
```diff
@@ -240,296 +162,443 @@ impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr = F>> Bat
...
+            Fr as BN382Fr,
+            Fq as BN382Fq,
+        };
+        use crate::{
+            BN382FrPoseidonHash, BN382FrBatchPoseidonHash,
+            BN382FqPoseidonHash, BN382FqBatchPoseidonHash,
+        };
+
+        #[test]
+        fn test_batch_hash_bn382fq() {
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/batched_crh.rs:439]: What's the use of batch evaluation for the BN382 x^5-Poseidon?

Comment (UlrichHaboeck75) [reply primitives/src/crh/poseidon/batched_crh.rs:439]: Ok, I have looked it up. For the BN382 we have the default implementation only, which just does a par.iter() over the batch.

File: primitives/src/crh/poseidon/mod.rs [hunk 25]
```diff
@@ -17,109 +21,105 @@ pub mod batched_crh;
...
+    }
 
-// Function that does the mix matrix
-#[allow(dead_code)]
-#[inline]
-pub fn matrix_mix<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr=F>>  (state: &mut Vec<F>) {
+    /// Perform scalar multiplication between vectors `res` and `state`,
+    /// modifying `res` in place.
+    #[inline]
+    fn scalar_mul(
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/mod.rs]: As mentioned in PR #69 : let us use a different name than scalar_mul.

File: primitives/src/crh/poseidon/mod.rs [hunk 26]
```diff
@@ -17,109 +21,105 @@ pub mod batched_crh;
...
+#[derivative(
+Clone(bound = ""),
+Debug(bound = ""),
+)]
+pub struct PoseidonHash<F: PrimeField, P: PoseidonParameters<Fr = F>, SB: SBox<Field = F, Parameters = P>>{
+    state: Vec<F>,
+    pending: Vec<F>,
+    input_size: Option<usize>,
+    updates_ctr: usize,
+    mod_rate: bool,
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/mod.rs:97]: As mentioned in PR #69 : we should explain the use of mod_rate explicitly.

File: primitives/src/crh/poseidon/mod.rs [hunk 27]
```diff
@@ -17,109 +21,105 @@ pub mod batched_crh;
...
 
-
         let mut instance = Self {
             state,
             pending: Vec::with_capacity(P::R),
             input_size: constant_size,
             updates_ctr: 0,
             mod_rate,
             _parameters: PhantomData,
+            _sbox: PhantomData,
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/mod.rs:122]: I addressed that in one of my comments on PR #69 : The personalization padding few lines below allows trivial collisions. We need to change to the 1,0,...0 padding as from the Poseidon paper.

Comment (UlrichHaboeck75) [reply primitives/src/crh/poseidon/mod.rs:122]: Will be changed in PR #69.

File: primitives/src/crh/poseidon/parameters/mnt4753.rs [hunk 28]
```diff
@@ -270,17 +290,13 @@ impl PoseidonParameters for MNT4753PoseidonParameters {
...
-        field_new!(Fr,BigInteger([0xa8a66953a924365d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0xb412c015510c2717, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x351fdbd63ac0afdb, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x302be8e2c8e27f02, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x7dcdc338f53308c, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x5220f8b41dab7db4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x524543d141024c82, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x3657a2432f363f4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-    ];
+    /// Short Montgomery multiplication with respect to the short Montgomery constant R_2=2^64
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/parameters/mnt4753.rs:293]: Sorry for giving the wrong instructions here: Let us write "Uses short Mongomery multiplication with respect to..."

File: primitives/src/crh/poseidon/parameters/mnt6753.rs [hunk 29]
```diff
@@ -271,17 +292,13 @@ impl PoseidonParameters for MNT6753PoseidonParameters {
...
-        field_new!(Fr,BigInteger([0xa8a66953a924365d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0xb412c015510c2717, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x351fdbd63ac0afdb, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x302be8e2c8e27f02, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x7dcdc338f53308c, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x5220f8b41dab7db4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x524543d141024c82, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-        field_new!(Fr,BigInteger([0x3657a2432f363f4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0])),
-    ];
+    /// Short Montgomery multiplication with respect to the short Montgomery constant R_2=2^64
```
Comment (UlrichHaboeck75) [inline primitives/src/crh/poseidon/parameters/mnt6753.rs:295]: As for the MNT4, let us write "Uses..."

File: primitives/src/merkle_tree/field_based_mht/parameters/bn382.rs [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+    biginteger::BigInteger384,
+    field_new,
+};
+
+use crate::{
+    crh::poseidon::BN382FrPoseidonHash,
+    FieldBasedMerkleTreePrecomputedEmptyConstants,
+};
+
+// PoseidonHash("This represents an empty Merkle Root for a BN382PoseidonHash based Merkle Tree.")
```
Comment (UlrichHaboeck75) [inline primitives/src/merkle_tree/field_based_mht/parameters/bn382.rs:12]: Let's add that this is for height 32 (?)

File: primitives/src/merkle_tree/field_based_mht/parameters/bn382_dual.rs [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+    biginteger::BigInteger384,
+    field_new,
+};
+
+use crate::{
+    crh::poseidon::BN382FqPoseidonHash,
+    FieldBasedMerkleTreePrecomputedEmptyConstants,
+};
+
+// PoseidonHash("This represents an empty Merkle Root for a BN382PoseidonHash based Merkle Tree.")
```
Comment (UlrichHaboeck75) [inline primitives/src/merkle_tree/field_based_mht/parameters/bn382_dual.rs:12]: Also here, let us add the height.

**Review by UlrichHaboeck75 — APPROVED** [review]

**Review by albertog78 — APPROVED** [review]

**Pull request merged**